enterprise/coderd/prebuilds/claim_test.go

```diff
@@ -0,0 +1,567 @@
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/enterprise/coderd/coderdenttest"
+	"github.com/coder/coder/v2/enterprise/coderd/prebuilds"
+	"github.com/coder/coder/v2/provisioner/echo"
+	"github.com/coder/coder/v2/provisionersdk/proto"
+	"github.com/coder/coder/v2/testutil"
+)
+
+type storeSpy struct {
+	database.Store
+
+	claims           *atomic.Int32
+	claimParams      *atomic.Pointer[database.ClaimPrebuiltWorkspaceParams]
+	claimedWorkspace *atomic.Pointer[database.ClaimPrebuiltWorkspaceRow]
+}
+
+func newStoreSpy(db database.Store) *storeSpy {
+	return &storeSpy{
+		Store:            db,
+		claims:           &atomic.Int32{},
+		claimParams:      &atomic.Pointer[database.ClaimPrebuiltWorkspaceParams]{},
+		claimedWorkspace: &atomic.Pointer[database.ClaimPrebuiltWorkspaceRow]{},
+	}
+}
+
+func (m *storeSpy) InTx(fn func(store database.Store) error, opts *database.TxOptions) error {
+	// Pass spy down into transaction store.
+	return m.Store.InTx(func(store database.Store) error {
+		spy := newStoreSpy(store)
+		spy.claims = m.claims
+		spy.claimParams = m.claimParams
+		spy.claimedWorkspace = m.claimedWorkspace
+
+		return fn(spy)
+	}, opts)
+}
+
+func (m *storeSpy) ClaimPrebuiltWorkspace(ctx context.Context, arg database.ClaimPrebuiltWorkspaceParams) (database.ClaimPrebuiltWorkspaceRow, error) {
+	m.claims.Add(1)
+	m.claimParams.Store(&arg)
+	result, err := m.Store.ClaimPrebuiltWorkspace(ctx, arg)
+	if err == nil {
+		m.claimedWorkspace.Store(&result)
+	}
+	return result, err
+}
+
+type errorStore struct {
+	claimingErr error
+
+	database.Store
+}
+
+func newErrorStore(db database.Store, claimingErr error) *errorStore {
+	return &errorStore{
+		Store:       db,
+		claimingErr: claimingErr,
+	}
+}
+
+func (es *errorStore) InTx(fn func(store database.Store) error, opts *database.TxOptions) error {
+	// Pass failure store down into transaction store.
+	return es.Store.InTx(func(store database.Store) error {
+		newES := newErrorStore(store, es.claimingErr)
+
+		return fn(newES)
+	}, opts)
+}
+
+func (es *errorStore) ClaimPrebuiltWorkspace(ctx context.Context, arg database.ClaimPrebuiltWorkspaceParams) (database.ClaimPrebuiltWorkspaceRow, error) {
+	return database.ClaimPrebuiltWorkspaceRow{}, es.claimingErr
+}
+
+func TestClaimPrebuild(t *testing.T) {
+	t.Parallel()
+
+	if !dbtestutil.WillUsePostgres() {
+		t.Skip("This test requires postgres")
+	}
+
+	const (
+		desiredInstances = 1
+		presetCount      = 2
+	)
+
+	cases := map[string]struct {
+		expectPrebuildClaimed  bool
+		markPrebuildsClaimable bool
+	}{
+		"no eligible prebuilds to claim": {
+			expectPrebuildClaimed:  false,
+			markPrebuildsClaimable: false,
+		},
+		"claiming an eligible prebuild should succeed": {
+			expectPrebuildClaimed:  true,
+			markPrebuildsClaimable: true,
+		},
+	}
+
+	for name, tc := range cases {
+		tc := tc
+
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			// Setup.
+			ctx := testutil.Context(t, testutil.WaitMedium)
```

Comment:
This may need to be bumped to `testutil.WaitLong`.

Reply:
Bumped to `testutil.WaitSuperLong` to be consistent with `require.Eventually()`.